arm_data_processing.c

```diff
@@ -27,12 +27,362 @@ Contact: Guillaume.Huard@imag.fr
+			shifter_carry_out = get_bit(Rm_value, 0);
+		} else { /* Rs[7:0] > 32 */
+			shifter_operand = 0;
+			shifter_carry_out = 0;
+		}
+	} else if (get_bits(shifter_operand_code, 6, 4) == 0b010) {
+		// lsr by immediate
+		if (shift_imm == 0) {
+			shifter_operand = 0;
+			shifter_carry_out = get_bit(Rm_value, 31);
+		} else { /* shift_imm > 0 */
+			shifter_operand = Rm_value >> shift_imm;
+			shifter_carry_out = get_bit(Rm_value, shift_imm - 1);
+		}
+	} else if (get_bits(shifter_operand_code, 7, 4) == 0b0011) {
+		// lsr by register
+		if (Rs_valueLB == 0) {
+			shifter_operand = Rm_value;
+			shifter_carry_out = C_bit;
+		} else if (Rs_valueLB < 32) {
+			shifter_operand = Rm_value >> Rs_valueLB;
+			shifter_carry_out = get_bit(Rm_value, Rs_valueLB - 1);
+		} else if (Rs_valueLB == 32) {
+			shifter_operand = 0;
+			shifter_carry_out = get_bit(Rm_value, 31);
+		} else { /* Rs[7:0] > 32 */
+			shifter_operand = 0;
+			shifter_carry_out = 0;
+		}
+	} else if (get_bits(shifter_operand_code, 6, 4) == 0b100) {
+		// asr by immediate
+		if (shift_imm == 0) {
+			if (get_bit(Rm_value, 31) == 0) {
+				shifter_operand = 0;
+				shifter_carry_out = 0;
+			} else { /* Rm[31] == 1 */
+				shifter_operand = 0xFFFFFFFF;
+				shifter_carry_out = 1;
+			}
+		} else { /* shift_imm > 0 */
+			shifter_operand = asr(Rm_value, shift_imm);
+			shifter_carry_out = get_bit(Rm_value, shift_imm - 1);
+		}
+	} else if (get_bits(shifter_operand_code, 7, 4) == 0b0101) {
+		// asr by register
+		if (Rs_valueLB == 0) {
+			shifter_operand = Rm_value;
+			shifter_carry_out = C_bit;
+		} else if (Rs_valueLB < 32) {
+			shifter_operand = asr(Rm_value, Rs_valueLB);
+			shifter_carry_out = get_bit(Rm_value, Rs_valueLB - 1);
+		} else { /* Rs_valueLB >= 32 */
+			if (get_bit(Rm_value, 31) == 0) {
+				shifter_operand = 0;
+				shifter_carry_out = 0;
+			} else { /* Rm[31] == 1 */
+				shifter_operand = 0xFFFFFFFF;
+				shifter_carry_out = 1;
+			}
+		}
+	} else if (get_bits(shifter_operand_code, 6, 4) == 0b110) {
+		// rotate right by immediate
+		if (shift_imm == 0) { /* rotate right with extend */
+			shifter_operand = ((uint32_t)C_bit << 31) | (Rm_value >> 1);
+			shifter_carry_out = get_bit(Rm_value, 0);
+		} else { /* shift_imm > 0 */
+			shifter_operand = ror(Rm_value, shift_imm);
+			shifter_carry_out = get_bit(Rm_value, shift_imm - 1);
+		}
+	} else if (get_bits(shifter_operand_code, 7, 4) == 0b0111) {
+		// rotate right by register
+		if (Rs_valueLB == 0) {
+			shifter_operand = Rm_value;
+			shifter_carry_out = C_bit;
+		} else if (get_bits(Rs_value, 4, 0) == 0) {
+			shifter_operand = Rm_value;
+			shifter_carry_out = get_bit(Rm_value, 31);
+		} else { /* Rs[4:0] > 0 */
+			shifter_operand = ror(Rm_value, get_bits(Rs_value, 4, 0));
+			shifter_carry_out = get_bit(Rm_value, get_bits(Rs_value, 4, 0) - 1);
+		}
+	} else
+		return UNDEFINED_INSTRUCTION;
+
+	/* CALCULATE RESULT BASED OPCODE */
+	switch (opcode) {
+	case AND:
+		result = Rn_value & shifter_operand;
```

Comment:
I skimmed through your code and the documentation, so I may be wrong, but shouldn't you check here whether rn (or rather before, when you extracted the value in bits) is equal to 15 and so you should add +8 to the address?
In the ARM documentation, it says:
Use of R15: If R15 is specified as register Rn, the value used is the address of the current instruction plus
eight.
Maybe you do this and I missed it...

Reply:
where did you saw this. I can only find when is with S, this case he treated

Reply:
ok found it. Yeah youssef is right i think there should be a case there to check if Rn is 15.

Reply:
Aight I will add it